### PR TITLE
[NetAppFiles] Apply 2020-12-01 api-version and recordings

### PR DESCRIPTION
**Pull request by audunn**
https://github.com/Azure/azure-rest-api-specs/pull/13388

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by audunn** [comment]
Note we had issues with generated validation code (not checking for null values) causes issues for SDK clients such as powershell. See: Azure/autorest#3585
However that issue seems to have been fixed and we can use that fix that issue by running autorest with
--use:@microsoft.azure/autorest.csharp@2.3.90 like so:

autorest.cmd https://github.com/Azure/azure-rest-api-specs/blob/master/specification/netapp/resource-manager/readme.md --csharp --version=v2 --use:@microsoft.azure/autorest.csharp@2.3.90 --reflect-api-versions --csharp-sdks-folder=C:\Users\****\Source\repos\azure\azure-sdk-for-net\sdk

Now it looks like this shows in the pipeline as it thinks generated code was changed. It was not changed in this case. The tool in the pipeline probably uses generate.ps1 that creates a slightly different autorest command line command that would include this validation bug (not checking for nulls).
Is there a way to update the tooling to use this autorest cmd line or resolve this pipeline issue some other way?

update: I tried to update the generate.ps1 script and ran locally with updated autorest command that worked. But it seems like the validation tool is using some other autorest command so that does not fix the issue.


This was also flagged in previous version https://github.com/Azure/azure-sdk-for-net/pull/18851

File: sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicy.cs [hunk 1]
```diff
@@ -40,22 +40,20 @@ public SnapshotPolicy()
...
         /// not</param>
         /// <param name="provisioningState">Azure lifecycle management</param>
-        public SnapshotPolicy(string location, string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), string name1 = default(string), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
+        public SnapshotPolicy(string location, string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
```
Comment (JonathanCrd) [inline sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicy.cs:50]: This is a breaking change: removing `string name1 = default(string)` param.

Has this breaking change been approved?

File: sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicyDetails.cs [hunk 2]
```diff
@@ -39,22 +39,20 @@ public SnapshotPolicyDetails()
...
         /// <param name="provisioningState">Azure lifecycle management</param>
-        public SnapshotPolicyDetails(string location = default(string), string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), string name1 = default(string), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
+        public SnapshotPolicyDetails(string location = default(string), string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
```
Comment (JonathanCrd) [inline sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicyDetails.cs:49]: Same here.

File: sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicyPatch.cs [hunk 3]
```diff
@@ -39,22 +39,20 @@ public SnapshotPolicyPatch()
...
         /// <param name="provisioningState">Azure lifecycle management</param>
-        public SnapshotPolicyPatch(string location = default(string), string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), string name1 = default(string), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
+        public SnapshotPolicyPatch(string location = default(string), string id = default(string), string name = default(string), string type = default(string), IDictionary<string, string> tags = default(IDictionary<string, string>), HourlySchedule hourlySchedule = default(HourlySchedule), DailySchedule dailySchedule = default(DailySchedule), WeeklySchedule weeklySchedule = default(WeeklySchedule), MonthlySchedule monthlySchedule = default(MonthlySchedule), bool? enabled = default(bool?), string provisioningState = default(string))
```
Comment (JonathanCrd) [inline sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicyPatch.cs:49]: Same here.

**Review by JonathanCrd — CHANGES_REQUESTED** [review]
I found out what it looks like a breaking change in SnapshotPolicy,
Everything else looks good, just CI tests still need to get fixed.

File: sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicy.cs (hunk 1 above)
Comment (audunn) [reply sdk/netapp/Microsoft.Azure.Management.NetApp/src/Generated/Models/SnapshotPolicy.cs:50]: We removed name property from snapshotpolicyProperties it is a violation of RPC .
https://armwiki.azurewebsites.net/api_contracts/RPCstrongvalidation.html

We have seen ICMs raised for this
`Resource response returned by provider 'Microsoft.NetApp' a JSON payload that contained invalid properties at the root scope, the invalid properties are 'Name'.`

https://portal.microsofticm.com/imp/v3/incidents/details/209491961/home

**Comment by audunn** [comment]
@JonathanCrd The CI issue is due to the pipeline incorrectly assumes discrepancy with generated code and reference code, it thinks generated code has been changed. It has not been changed and this was an issue with last release as well due to AutoRest flags being used.

@allenjzhang can you assist with this issue?

**Comment by JonathanCrd** [comment]
> @JonathanCrd Jonathan Cardenas FTE The CI issue is due to the pipeline incorrectly assumes discrepancy with generated code and reference code, it thinks generated code has been changed. It has not been changed and this was an issue with last release as well due to AutoRest flags being used.
> 
> @allenjzhang Allen Zhang (DEVDIV) FTE can you assist with this issue?

Yes, the reason they are failing it's because the version is hardcoded [in the generate script](https://github.com/Azure/azure-sdk-for-net/blob/26353c8628b2145e9dd8ccd8a22bfe841a377795/sdk/netapp/Microsoft.Azure.Management.NetApp/src/generate.ps1).

Everything looks good, so I'm going to merge this PR.

**Review by JonathanCrd — APPROVED** [review]

**Comment by audunn** [comment]
@JonathanCrd   I don't know of any other way to get around the original issue. The version of AutoRest used was generating faulty validation code, by not performing null-check before trying to validate nullable properties using the % operator  https://github.com/Azure/autorest/issues/3585

**Pull request merged**